File: Typescript/2021/02/day2_part2.ts
export class Submarine {
  #position: number;
  #depth: number;
  #aim: number;

  constructor() {
    this.#position = 0;
    this.#depth = 0;
    this.#aim = 0;
  }
  get position() {
    return this.#position;
  }
  get depth() {
    return this.#depth;
  }
  get aim() {
    return this.#aim;
  }
  takeOrders(listOfCommands: string[]) {
    listOfCommands.forEach((command) => {
      this.move(command);
    });
  }
  move(input: string) {
    const [command, n] = input.split(" ");
    switch (command) {
      case "forward":
        this.forward(parseInt(n));
        break;
      case "down":
        this.down(parseInt(n));
        break;
      case "up":
        this.up(parseInt(n));
        break;
    }
  }
  forward(n: number) {
    this.#position += n;
    this.#depth += this.#aim * n;
  }
  down(n: number) {
    this.#aim += n;
  }
  up(n: number) {
    this.#aim -= n;
  }
}

File: Python/day2/day2_part2.py
class Submarine():
  def __init__(self) -> None:
      self.depth = 0
      self.horisontal = 0
      self.aim = 0

  def change_horisontal(self,n):
    self.horisontal += n
    self.depth += self.aim*n
  
  def change_depth(self,n):
    self.depth += n

  def change_aim(self,n):
    self.aim += n
  
  def parse_moves(self,moves):
    moves = [(x.split(" ")[0],int(x.split(" ")[1])) for x in moves]
    for move in moves:
      if move[0] == "up":
        self.change_aim(-move[1])
      if move[0] == "down":
        self.change_aim(move[1])
      if move[0] == "forward":
        self.change_horisontal(move[1])
    
  def print_position(self):
    print("Depth: ",self.depth)
    print("Horisontal:",self.horisontal)
    print("Combined",self.depth*self.horisontal)


moves = []
with open("day2.txt","r") as f:
  moves = f.readlines()


sub = Submarine()
sub.parse_moves(moves)
sub.print_position()
